tests/test_utils.py

```diff
@@ -1819,6 +1819,242 @@ def test_set_index_dtypes(index, dtypes, expected):
+                [0, 0, 0],
+                [1, 1, pd.NaT],
+            ),
+        ),
+        (
+            [
+                audformat.segmented_index(['f1', 'f2'], [0, 0], [1, 1]),
+                audformat.filewise_index(['f1', 'f2']),
+                audformat.filewise_index(['f2', 'f3']),
+            ],
+            audformat.segmented_index(
+                ['f1', 'f2', 'f1', 'f3'],
+                [0, 0, 0, 0],
+                [1, 1, pd.NaT, pd.NaT],
+            ),
+        ),
+        (
+            [
+                audformat.filewise_index(['f1', 'f2']),
+                audformat.segmented_index(['f1', 'f2'], [0, 0], [1, 1]),
+                audformat.filewise_index(['f2', 'f3']),
+            ],
+            audformat.segmented_index(
+                ['f1', 'f2', 'f1', 'f3'],
+                [0, 0, 0, 0],
+                [1, 1, pd.NaT, pd.NaT],
+            ),
+        ),
+        # other
+        (
+            [
+                pd.Index([]),
+            ],
+            pd.Index([]),
+        ),
+        (
+            [
+                pd.Index([]),
+                pd.Index([]),
+            ],
+            pd.Index([]),
+        ),
+        (
+            [
+                pd.Index([0, 1], name='idx'),
+                pd.Index([1, 2], name='idx'),
+            ],
+            pd.Index([0, 2], name='idx'),
+        ),
+        (
+            [
+                pd.Index([0, 1], name='idx'),
+                pd.Index([1, 2], dtype='Int64', name='idx'),
+            ],
+            pd.Index([0, 2], dtype='Int64', name='idx'),
+        ),
+        (
+            [
+                pd.Index([0, 1], name='idx'),
+                pd.MultiIndex.from_arrays([[1, 2]], names=['idx']),
+            ],
+            pd.Index([0, 2], name='idx'),
+        ),
+        (
+            [
+                pd.MultiIndex.from_arrays([[0, 1]], names=['idx']),
+                pd.MultiIndex.from_arrays([[1, 2]], names=['idx']),
+            ],
+            pd.MultiIndex.from_arrays([[0, 2]], names=['idx']),
+        ),
+        (
+            [
+                pd.MultiIndex.from_arrays(
+                    [['a', 'b', 'c'], [0, 1, 2]],
+                    names=['idx1', 'idx2'],
+                ),
+                pd.MultiIndex.from_arrays(
+                    [['b', 'c'], [1, 3]],
+                    names=['idx1', 'idx2'],
+                ),
+            ],
+            pd.MultiIndex.from_arrays(
+                [['a', 'c', 'c'], [0, 2, 3]],
+                names=['idx1', 'idx2'],
+            ),
+        ),
+        pytest.param(
+            [
+                pd.Index([], name='idx1'),
+                pd.Index([], name='idx2'),
+            ],
+            None,
+            marks=pytest.mark.xfail(raises=ValueError),
+        ),
+    ]
+)
+def test_symmetric_difference(objs, expected):
+    pd.testing.assert_index_equal(
+        audformat.utils.symmetric_difference(objs),
+        expected,
+    )
+
```

Comment:
I'm not completely sure yet if (A Δ B) Δ C = A Δ (B Δ C) should also hold with our special filewise/segmented index handling.
But if yes, you could add a test for it here with:
```suggestion
    if len(objs) > 2: 
        pd.testing.assert_index_equal(
            audformat.utils.symmetric_difference(
                [
                    objs[0],
                    audformat.utils.symmetric_difference(objs[1:]),
                ]
            ),
            audformat.utils.symmetric_difference(
                [
                    audformat.utils.symmetric_difference(objs[:-1]),
                    objs[-1],
                ]
            ),
        )

```

Reply:
As you already assumed it fails when a combination of filewise and segmented index objects is passed.

Reply:
Yes, but it still looks like a bug in the implementation.

The first `objs` it fails for is:
```python
index1 = audformat.segmented_index(['f1', 'f2'], [0, 0], [1, 1])
index2 = audformat.segmented_index(['f2', 'f3'], [0, 0], [1, 1])
index3 = audformat.filewise_index(['f1', 'f2'])
objs = [
    index1,
    index2,
    index3,
]
```
The result is:
```python
audformat.utils.symmetric_difference([index1, index2, index3])
```
```
MultiIndex([('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```
You get the same when doing:
```python
audformat.utils.symmetric_difference(
    [
        audformat.utils.symmetric_difference([index1, index2]),
        index3,
    ]
)
```
```
MultiIndex([('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```
but a different result when doing:
```python
audformat.utils.symmetric_difference(
    [
        index1,
        audformat.utils.symmetric_difference([index2, index3]),
    ]
)
```
```
MultiIndex([('f1', '0 days', '0 days 00:00:01'),
            ('f2', '0 days', '0 days 00:00:01'),
            ('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```

---

But the last result is wrong in my opinion.
You have
```python
index1
```
```
MultiIndex([('f1', '0 days', '0 days 00:00:01'),
            ('f2', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```
and
```python
audformat.utils.symmetric_difference([index2, index3])
```
```
MultiIndex([('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```
if you take the symmetric difference between them it should be
```
MultiIndex([('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```
as we said we remove segments if the whole file is present.

Reply:
It's because when you do:

```python
audformat.utils.symmetric_difference(
    [
        index1,
        audformat.utils.symmetric_difference([index2, index3]),
    ]
)
```

you are passing only segmented index objects.

Reply:
I agree it's not very nice that we have a different behavior now whether you pass a mixture of filewise and segmented vs. only segmented index object. But then we would need to implement a function that removes all segments that lie within another segment first.

Reply:
The same issue we have with `intersect)` btw:

```python
objs = [
    audformat.filewise_index(['f1', 'f2']),
    audformat.segmented_index('f1', 0, 1),
]
audformat.utils.intersect(objs)
```
```
MultiIndex([('f1', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```

vs.

```python
objs = [
    audformat.segmented_index(['f1', 'f2']),
    audformat.segmented_index('f1', 0, 1),
]
audformat.utils.intersect(objs)
```
```
MultiIndex([('f1', '0 days', 'NaT')],
           names=['file', 'start', 'end'])
```

Here we would require a utility function that removes segments that include other segments.

Reply:
I think it would be worse the effort as from a user perspective there is not really a difference between
```python
audformat.filewise_index([f1'])
```
and
```python
audformat.segmented_index(['f1'])
```

Reply:
Ok, should we then stop here and try to finish #225 without this PR?

Reply:
Basically we need to convert back to filewise if we have a `NaT`. We have `audformat.utils.to_filewise_index()`, but this is doing something completely different ;)

Reply:
On the other hand - this might get quite complicated, e.g.:

```python
objs = [
    audformat.segmented_index('f1', 0, 2),
    audformat.segmented_index('f1', 1, 3),
]
audformat.utils.intersect(objs)
```

This currently returns:

```
MultiIndex([], names=['file', 'start', 'end'])
```

but correct would be:

```
MultiIndex([('f1', '0 days 00:00:01', '0 days 00:00:02')],
           names=['file', 'start', 'end'])
```

Reply:
But there might be so many special cases and stuff, I wonder if it is really worth the effort...

Reply:
My opinion: continue as as is and support only a mixture of filewise + segmented as special case.

Reply:
But what indeed seems to be a bug is:

```python
objs = [
    audformat.segmented_index('f1'),
    audformat.segmented_index('f1', 0, 1),
]
audformat.utils.intersect(objs)
```
```
MultiIndex([('f1', '0 days', NaT)],
           names=['file', 'start', 'end'])
```

This should return:

```
MultiIndex([], names=['file', 'start', 'end'])
```

Reply:
The other way round it works:

```python
objs = [
    audformat.segmented_index('f1', 0, 1),
    audformat.segmented_index('f1'),
]
audformat.utils.intersect(objs)
```
```
MultiIndex([], names=['file', 'start', 'end'])
```

Reply:
Sounds like we should add tests for switching the order of `objs` for our utility functions.

Reply:
I think it's a problem with `pd.NaT` and that it matches anything.

Reply:
> My opinion: continue as as is and support only a mixture of filewise + segmented as special case.

In order to get the first release out with `MiscTable` this sounds reasonable. But I would open an issue, stating that we should fix this before we make the 1.0.0 release.

Reply:
Because this works:

```python
objs = [
    audformat.segmented_index('f1', 0, 1),
    audformat.segmented_index('f1', 0, 2),
]
audformat.utils.intersect(objs)
```
```
MultiIndex([], names=['file', 'start', 'end'])
```

and:


```python
objs = [
    audformat.segmented_index('f1', 0, 2),
    audformat.segmented_index('f1', 0, 1),
]
audformat.utils.intersect(objs)
```
```
MultiIndex([], names=['file', 'start', 'end'])
```

Reply:
> In order to get the first release out with MiscTable this sounds reasonable. But I would open an issue, stating that we should fix this before we make the 1.0.0 release.

Yes, but I would only fix the order bug, whereas the following seems correct to me:

```python
objs = [
    audformat.segmented_index('f1', 0, 1),
    audformat.segmented_index('f1'),
]
audformat.utils.intersect(objs)
```
```
MultiIndex([], names=['file', 'start', 'end'])
```

Cause let's say you have a segmented index and it includes `'f1', 0, 1` and `'f1', 0, pd.NaT`  and then you drop `'f1', 0, pd.NaT`, I don't think you necessarily expect that also `'f1', 0, 1` is dropped. If you a user really wants this, then she can do it by passing a filewise index to explicitly force that all segments belonging to `'f1'` should be removed.

Reply:
The correct example would be:

```python
index = audformat.segmented_index(['f1', 'f1'], [0, 0], [1, pd.NaT])
audformat.utils.intersect(
    [
        index,
        audformat.segmented_index('f1'),
    ]
)
```
```
MultiIndex([('f1', '0 days', 'NaT')],
           names=['file', 'start', 'end'])
```

vs.

``` python
index = audformat.segmented_index(['f1', 'f1'], [0, 0], [1, pd.NaT])
audformat.utils.intersect(
    [
        index,
        audformat.filewise_index('f1'),
    ]
)
```
```
MultiIndex([('f1', '0 days', '0 days 00:00:01'),
            ('f1', '0 days',               NaT)],
           names=['file', 'start', 'end'])
```

Reply:
Which means you would argue the behavior I described above is not a bug, but should be the expected behavior for `symmetric_difference()`, correct?

Reply:
I would also propose to stay with your current implementation here, the question is only we should open a bug and require (A Δ B) Δ C = A Δ (B Δ C). And if not, if we should mention somewhere in the docstrings that this cannot be assumed for a mixture of filewise and segmented indices.

Reply:
> ![image](https://user-images.githubusercontent.com/173624/181281548-1258bd22-9f32-4c3c-8500-5d716c475924.png)

No, we agreed already that we will not handle such overlapping cases. Only the case of `NaT` is considered.
The correct return value is indeed what is returned from the function at the moment.